File: Training/Checkpointing/pi_checkpoint.py
#!/usr/bin/env python3

"""
pi_checkpoint.py

Estimate pi using a Monte Carlo "dart throwing" experiment with
application-level checkpoint/restart.

The checkpoint contains all state required to continue the calculation:

    - number of completed darts
    - number of darts inside the unit circle
    - random-number generator state
    - elapsed runtime
    - run configuration

Examples
--------
Start a new calculation:

    python pi_checkpoint.py \
        --darts 100000000 \
        --seed 42 \
        --report-every 1000000 \
        --checkpoint-every 1000000 \
        --sleep 1

Restart from a checkpoint:

    python pi_checkpoint.py \
        --darts 100000000 \
        --seed 42 \
        --report-every 1000000 \
        --checkpoint-every 1000000 \
        --sleep 1 \
        --resume
"""

import argparse
import math
import os
import pickle
import random
import time
from pathlib import Path


CHECKPOINT_VERSION = 1


def parse_args():
    parser = argparse.ArgumentParser(
        description="Estimate pi using Monte Carlo dart throwing "
                    "with checkpoint/restart."
    )

    parser.add_argument(
        "-n",
        "--darts",
        type=int,
        default=1_000_000,
        help="Number of darts to throw (default: 1,000,000)",
    )

    parser.add_argument(
        "--seed",
        type=int,
        default=42,
        help="Random-number generator seed (default: 42)",
    )

    parser.add_argument(
        "--report-every",
        type=int,
        default=100_000,
        help="Print progress every N darts (default: 100,000)",
    )

    parser.add_argument(
        "--checkpoint-every",
        type=int,
        default=100_000,
        help="Write a checkpoint every N darts (default: 100,000)",
    )

    parser.add_argument(
        "--checkpoint-file",
        type=Path,
        default=Path("pi_checkpoint.pkl"),
        help="Checkpoint filename (default: pi_checkpoint.pkl)",
    )

    parser.add_argument(
        "--resume",
        action="store_true",
        help="Resume calculation from an existing checkpoint",
    )

    parser.add_argument(
        "--sleep",
        type=float,
        default=0.0,
        help="Optional delay after each progress report "
             "(useful for demonstrations)",
    )

    return parser.parse_args()


def save_checkpoint(
    checkpoint_file,
    completed_darts,
    inside_circle,
    rng,
    n_darts,
    seed,
    elapsed,
):
    """
    Save checkpoint atomically.

    The new checkpoint is first written to a temporary file. Only after
    the write succeeds is it atomically renamed to the real checkpoint
    filename.
    """

    checkpoint = {
        "version": CHECKPOINT_VERSION,
        "completed_darts": completed_darts,
        "inside_circle": inside_circle,
        "rng_state": rng.getstate(),
        "n_darts": n_darts,
        "seed": seed,
        "elapsed": elapsed,
    }

    checkpoint_file.parent.mkdir(parents=True, exist_ok=True)

    tmp_file = checkpoint_file.with_name(
        checkpoint_file.name + ".tmp"
    )

    with open(tmp_file, "wb") as f:
        pickle.dump(
            checkpoint,
            f,
            protocol=pickle.HIGHEST_PROTOCOL,
        )

        # Push Python's buffered data to the operating system.
        f.flush()

        # Request that the operating system flush the file to storage.
        os.fsync(f.fileno())

    # Atomic replacement on the same filesystem.
    os.replace(tmp_file, checkpoint_file)


def load_checkpoint(checkpoint_file, expected_darts, expected_seed):
    """
    Load and validate an existing checkpoint.
    """

    if not checkpoint_file.exists():
        raise FileNotFoundError(
            f"Checkpoint file '{checkpoint_file}' does not exist."
        )

    with open(checkpoint_file, "rb") as f:
        checkpoint = pickle.load(f)

    if checkpoint.get("version") != CHECKPOINT_VERSION:
        raise RuntimeError(
            "Checkpoint version is incompatible with this program."
        )

    if checkpoint["n_darts"] != expected_darts:
        raise RuntimeError(
            "Checkpoint was created for a different number of darts:\n"
            f"  checkpoint : {checkpoint['n_darts']:,}\n"
            f"  requested  : {expected_darts:,}"
        )

    if checkpoint["seed"] != expected_seed:
        raise RuntimeError(
            "Checkpoint was created with a different random seed:\n"
            f"  checkpoint : {checkpoint['seed']}\n"
            f"  requested  : {expected_seed}"
        )

    return checkpoint


def estimate_pi(args):
    """
    Run or resume the Monte Carlo calculation.
    """

    rng = random.Random(args.seed)

    completed_darts = 0
    inside_circle = 0
    previous_elapsed = 0.0

    #
    # Restart if requested.
    #
    if args.resume:
        checkpoint = load_checkpoint(
            checkpoint_file=args.checkpoint_file,
            expected_darts=args.darts,
            expected_seed=args.seed,
        )

        completed_darts = checkpoint["completed_darts"]
        inside_circle = checkpoint["inside_circle"]
        previous_elapsed = checkpoint["elapsed"]

        # Restore the exact state of the random-number generator.
        rng.setstate(checkpoint["rng_state"])

        print()
        print("Checkpoint found.")
        print(f"Resuming from dart : {completed_darts:,}")
        print(f"Inside circle      : {inside_circle:,}")
        print(f"Previous runtime   : {previous_elapsed:.2f} s")
        print()

    start_time = time.perf_counter()

    #
    # Continue from the first dart not already completed.
    #
    for i in range(completed_darts + 1, args.darts + 1):

        x = rng.uniform(-1.0, 1.0)
        y = rng.uniform(-1.0, 1.0)

        if x * x + y * y <= 1.0:
            inside_circle += 1

        #
        # Progress report
        #
        if args.report_every > 0 and (
            i % args.report_every == 0 or i == args.darts
        ):
            pi_estimate = 4.0 * inside_circle / i

            elapsed = (
                previous_elapsed
                + time.perf_counter()
                - start_time
            )

            print(
                f"darts = {i:>12,d} / {args.darts:,}   "
                f"pi = {pi_estimate:.8f}   "
                f"error = {abs(pi_estimate - math.pi):.3e}   "
                f"elapsed = {elapsed:.2f} s",
                flush=True,
            )

            if args.sleep > 0:
                time.sleep(args.sleep)

        #
        # Periodic checkpoint
        #
        if args.checkpoint_every > 0 and (
            i % args.checkpoint_every == 0
        ):
            elapsed = (
                previous_elapsed
                + time.perf_counter()
                - start_time
            )

            save_checkpoint(
                checkpoint_file=args.checkpoint_file,
                completed_darts=i,
                inside_circle=inside_circle,
                rng=rng,
                n_darts=args.darts,
                seed=args.seed,
                elapsed=elapsed,
            )

            print(
                f"  --> checkpoint saved at dart {i:,}",
                flush=True,
            )

    total_elapsed = (
        previous_elapsed
        + time.perf_counter()
        - start_time
    )

    pi_estimate = 4.0 * inside_circle / args.darts

    #
    # Write one final checkpoint representing the completed calculation.
    #
    save_checkpoint(
        checkpoint_file=args.checkpoint_file,
        completed_darts=args.darts,
        inside_circle=inside_circle,
        rng=rng,
        n_darts=args.darts,
        seed=args.seed,
        elapsed=total_elapsed,
    )

    return pi_estimate, inside_circle, total_elapsed


def main():
    args = parse_args()

    if args.darts <= 0:
        raise ValueError("--darts must be greater than zero")

    if args.report_every < 0:
        raise ValueError("--report-every cannot be negative")

    if args.checkpoint_every < 0:
        raise ValueError("--checkpoint-every cannot be negative")

    if args.sleep < 0:
        raise ValueError("--sleep cannot be negative")

    print("=" * 72)
    print("Monte Carlo Pi Estimation with Checkpoint/Restart")
    print("=" * 72)
    print(f"Number of darts      : {args.darts:,}")
    print(f"Random seed          : {args.seed}")
    print(f"Report interval      : {args.report_every:,}")
    print(f"Checkpoint interval  : {args.checkpoint_every:,}")
    print(f"Checkpoint file      : {args.checkpoint_file}")
    print(f"Resume               : {args.resume}")
    print(f"Sleep interval       : {args.sleep} s")
    print("=" * 72)

    pi_estimate, inside_circle, elapsed = estimate_pi(args)

    print()
    print("=" * 72)
    print("Final result")
    print("=" * 72)
    print(f"Darts thrown       : {args.darts:,}")
    print(f"Inside circle      : {inside_circle:,}")
    print(f"Estimated pi       : {pi_estimate:.10f}")
    print(f"Python math.pi     : {math.pi:.10f}")
    print(f"Absolute error     : {abs(pi_estimate - math.pi):.6e}")
    print(f"Total elapsed time : {elapsed:.2f} s")
    print("=" * 72)


if __name__ == "__main__":
    main()

File: Training/Checkpointing/Exercise2/pi_checkpoint.c
/*
 * pi_checkpoint.c
 *
 * Monte Carlo estimation of pi with application-level
 * checkpoint/restart.
 *
 * Demonstrates:
 *
 *   - explicit application state
 *   - binary checkpoint files
 *   - RNG-state preservation
 *   - temporary-file + atomic rename
 *   - manual restart using --resume
 */

#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <math.h>


#define CHECKPOINT_VERSION 1

static const double PI_TRUE =
    3.141592653589793238462643383279502884;


/*
 * ------------------------------------------------------------------
 * Checkpoint state
 * ------------------------------------------------------------------
 */

typedef struct {

    char magic[8];

    uint32_t version;

    uint64_t completed_darts;
    uint64_t inside_circle;

    uint64_t rng_state;

    uint64_t total_darts;
    uint64_t seed;

    double elapsed_seconds;

} checkpoint_t;


/*
 * Runtime options.
 */

typedef struct {

    uint64_t darts;
    uint64_t seed;

    uint64_t report_every;
    uint64_t checkpoint_every;

    double sleep_seconds;

    const char *checkpoint_file;

    int resume;

} options_t;


/*
 * ------------------------------------------------------------------
 * RNG
 * ------------------------------------------------------------------
 */

static uint64_t rng_next(uint64_t *state)
{
    uint64_t x = *state;

    x ^= x >> 12;
    x ^= x << 25;
    x ^= x >> 27;

    *state = x;

    return x * UINT64_C(2685821657736338717);
}


static double rng_uniform(uint64_t *state)
{
    return (rng_next(state) >> 11) *
           (1.0 / 9007199254740992.0);
}


/*
 * ------------------------------------------------------------------
 * Timing
 * ------------------------------------------------------------------
 */

static double wall_time(void)
{
    struct timespec ts;

    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
        perror("clock_gettime");
        exit(EXIT_FAILURE);
    }

    return (double) ts.tv_sec +
           (double) ts.tv_nsec * 1.0e-9;
}


static void sleep_seconds(double seconds)
{
    if (seconds <= 0.0) {
        return;
    }

    struct timespec req;

    req.tv_sec = (time_t) seconds;
    req.tv_nsec =
        (long) ((seconds - req.tv_sec) * 1.0e9);

    while (nanosleep(&req, &req) != 0) {

        if (errno != EINTR) {
            perror("nanosleep");
            exit(EXIT_FAILURE);
        }
    }
}


/*
 * ------------------------------------------------------------------
 * Checkpoint I/O
 * ------------------------------------------------------------------
 */

static void initialize_checkpoint(
    checkpoint_t *checkpoint
)
{
    memset(checkpoint, 0, sizeof(*checkpoint));

    memcpy(
        checkpoint->magic,
        "PICHECK",
        7
    );

    checkpoint->version =
        CHECKPOINT_VERSION;
}


static int checkpoint_is_valid(
    const checkpoint_t *checkpoint
)
{
    if (
        memcmp(
            checkpoint->magic,
            "PICHECK",
            7
        ) != 0
    ) {
        return 0;
    }

    if (
        checkpoint->version !=
        CHECKPOINT_VERSION
    ) {
        return 0;
    }

    return 1;
}


static void save_checkpoint(
    const char *filename,
    const checkpoint_t *checkpoint
)
{
    char tmp_filename[4096];

    int written = snprintf(
        tmp_filename,
        sizeof(tmp_filename),
        "%s.tmp",
        filename
    );

    if (
        written < 0 ||
        (size_t) written >= sizeof(tmp_filename)
    ) {
        fprintf(
            stderr,
            "Checkpoint filename is too long.\n"
        );
        exit(EXIT_FAILURE);
    }


    FILE *file =
        fopen(tmp_filename, "wb");

    if (file == NULL) {
        perror("fopen checkpoint");
        exit(EXIT_FAILURE);
    }


    if (
        fwrite(
            checkpoint,
            sizeof(*checkpoint),
            1,
            file
        ) != 1
    ) {

        perror("fwrite checkpoint");
        fclose(file);
        exit(EXIT_FAILURE);
    }


    /*
     * Flush C stdio buffers.
     */
    if (fflush(file) != 0) {
        perror("fflush checkpoint");
        fclose(file);
        exit(EXIT_FAILURE);
    }


    /*
     * Ask the OS to flush the underlying file.
     */
    if (fsync(fileno(file)) != 0) {
        perror("fsync checkpoint");
        fclose(file);
        exit(EXIT_FAILURE);
    }


    if (fclose(file) != 0) {
        perror("fclose checkpoint");
        exit(EXIT_FAILURE);
    }


    /*
     * Atomic replacement on the same filesystem.
     */
    if (
        rename(
            tmp_filename,
            filename
        ) != 0
    ) {
        perror("rename checkpoint");
        exit(EXIT_FAILURE);
    }
}


static checkpoint_t load_checkpoint(
    const char *filename
)
{
    checkpoint_t checkpoint;

    FILE *file =
        fopen(filename, "rb");

    if (file == NULL) {
        perror("fopen checkpoint");
        exit(EXIT_FAILURE);
    }


    if (
        fread(
            &checkpoint,
            sizeof(checkpoint),
            1,
            file
        ) != 1
    ) {

        fprintf(
            stderr,
            "Failed to read checkpoint file.\n"
        );

        fclose(file);
        exit(EXIT_FAILURE);
    }


    if (fclose(file) != 0) {
        perror("fclose checkpoint");
        exit(EXIT_FAILURE);
    }


    if (!checkpoint_is_valid(&checkpoint)) {

        fprintf(
            stderr,
            "Invalid or incompatible "
            "checkpoint file.\n"
        );

        exit(EXIT_FAILURE);
    }


    return checkpoint;
}


/*
 * ------------------------------------------------------------------
 * CLI helpers
 * ------------------------------------------------------------------
 */

static void usage(const char *program)
{
    printf(
        "Usage: %s [OPTIONS]\n\n"
        "Options:\n"
        "  -n, --darts N\n"
        "      --seed N\n"
        "      --report-every N\n"
        "      --checkpoint-every N\n"
        "      --checkpoint-file FILE\n"
        "      --sleep SEC\n"
        "      --resume\n"
        "  -h, --help\n",
        program
    );
}


static uint64_t parse_uint64(
    const char *text,
    const char *name
)
{
    char *end = NULL;

    errno = 0;

    unsigned long long value =
        strtoull(text, &end, 10);

    if (
        errno != 0 ||
        end == text ||
        *end != '\0'
    ) {

        fprintf(
            stderr,
            "Invalid value for %s: %s\n",
            name,
            text
        );

        exit(EXIT_FAILURE);
    }

    return (uint64_t) value;
}


static double parse_double(
    const char *text,
    const char *name
)
{
    char *end = NULL;

    errno = 0;

    double value =
        strtod(text, &end);

    if (
        errno != 0 ||
        end == text ||
        *end != '\0'
    ) {

        fprintf(
            stderr,
            "Invalid value for %s: %s\n",
            name,
            text
        );

        exit(EXIT_FAILURE);
    }

    return value;
}


static options_t parse_args(
    int argc,
    char **argv
)
{
    options_t options = {
        .darts = UINT64_C(1000000),
        .seed = UINT64_C(42),
        .report_every = UINT64_C(100000),
        .checkpoint_every = UINT64_C(100000),
        .sleep_seconds = 0.0,
        .checkpoint_file = "pi_checkpoint.bin",
        .resume = 0
    };


    enum {
        OPT_SEED = 1000,
        OPT_REPORT,
        OPT_CHECKPOINT_EVERY,
        OPT_CHECKPOINT_FILE,
        OPT_SLEEP,
        OPT_RESUME
    };


    static const struct option long_options[] = {

        {"darts", required_argument, NULL, 'n'},

        {"seed",
         required_argument,
         NULL,
         OPT_SEED},

        {"report-every",
         required_argument,
         NULL,
         OPT_REPORT},

        {"checkpoint-every",
         required_argument,
         NULL,
         OPT_CHECKPOINT_EVERY},

        {"checkpoint-file",
         required_argument,
         NULL,
         OPT_CHECKPOINT_FILE},

        {"sleep",
         required_argument,
         NULL,
         OPT_SLEEP},

        {"resume",
         no_argument,
         NULL,
         OPT_RESUME},

        {"help",
         no_argument,
         NULL,
         'h'},

        {NULL, 0, NULL, 0}
    };


    int option;

    while (
        (option = getopt_long(
            argc,
            argv,
            "n:h",
            long_options,
            NULL
        )) != -1
    ) {

        switch (option) {

            case 'n':
                options.darts =
                    parse_uint64(
                        optarg,
                        "--darts"
                    );
                break;

            case OPT_SEED:
                options.seed =
                    parse_uint64(
                        optarg,
                        "--seed"
                    );
                break;

            case OPT_REPORT:
                options.report_every =
                    parse_uint64(
                        optarg,
                        "--report-every"
                    );
                break;

            case OPT_CHECKPOINT_EVERY:
                options.checkpoint_every =
                    parse_uint64(
                        optarg,
                        "--checkpoint-every"
                    );
                break;

            case OPT_CHECKPOINT_FILE:
                options.checkpoint_file =
                    optarg;
                break;

            case OPT_SLEEP:
                options.sleep_seconds =
                    parse_double(
                        optarg,
                        "--sleep"
                    );
                break;

            case OPT_RESUME:
                options.resume = 1;
                break;

            case 'h':
                usage(argv[0]);
                exit(EXIT_SUCCESS);

            default:
                usage(argv[0]);
                exit(EXIT_FAILURE);
        }
    }


    if (options.darts == 0) {
        fprintf(
            stderr,
            "--darts must be > 0\n"
        );
        exit(EXIT_FAILURE);
    }


    if (options.seed == 0) {
        fprintf(
            stderr,
            "--seed must be nonzero\n"
        );
        exit(EXIT_FAILURE);
    }


    if (options.sleep_seconds < 0.0) {
        fprintf(
            stderr,
            "--sleep cannot be negative\n"
        );
        exit(EXIT_FAILURE);
    }


    return options;
}


/*
 * ------------------------------------------------------------------
 * Main
 * ------------------------------------------------------------------
 */

int main(int argc, char **argv)
{
    options_t options =
        parse_args(argc, argv);


    uint64_t completed_darts = 0;
    uint64_t inside_circle = 0;

    uint64_t rng_state =
        options.seed;

    double previous_elapsed = 0.0;


    /*
     * --------------------------------------------------------------
     * Restore checkpoint
     * --------------------------------------------------------------
     */

    if (options.resume) {

        checkpoint_t checkpoint =
            load_checkpoint(
                options.checkpoint_file
            );


        if (
            checkpoint.total_darts !=
            options.darts
        ) {

            fprintf(
                stderr,
                "Checkpoint belongs to a "
                "different calculation.\n"
            );

            exit(EXIT_FAILURE);
        }


        if (
            checkpoint.seed !=
            options.seed
        ) {

            fprintf(
                stderr,
                "Checkpoint uses a "
                "different seed.\n"
            );

            exit(EXIT_FAILURE);
        }


        completed_darts =
            checkpoint.completed_darts;

        inside_circle =
            checkpoint.inside_circle;

        rng_state =
            checkpoint.rng_state;

        previous_elapsed =
            checkpoint.elapsed_seconds;


        printf("\n");
        printf("========================================================================\n");
        printf("Restarting from checkpoint\n");
        printf("========================================================================\n");

        printf(
            "Completed darts  : %" PRIu64 "\n",
            completed_darts
        );

        printf(
            "Inside circle    : %" PRIu64 "\n",
            inside_circle
        );

        printf(
            "Previous runtime : %.2f s\n",
            previous_elapsed
        );

        printf("========================================================================\n\n");
    }


    const double start_time =
        wall_time();


    /*
     * --------------------------------------------------------------
     * Main loop
     * --------------------------------------------------------------
     */

    for (
        uint64_t i = completed_darts + 1;
        i <= options.darts;
        ++i
    ) {

        const double x =
            2.0 * rng_uniform(&rng_state)
            - 1.0;

        const double y =
            2.0 * rng_uniform(&rng_state)
            - 1.0;


        if (x * x + y * y <= 1.0) {
            ++inside_circle;
        }


        /*
         * Progress report
         */

        if (
            options.report_every > 0 &&
            (
                i % options.report_every == 0 ||
                i == options.darts
            )
        ) {

            const double pi_estimate =
                4.0 *
                (double) inside_circle /
                (double) i;


            const double elapsed =
                previous_elapsed +
                wall_time() -
                start_time;


            printf(
                "darts = %12" PRIu64
                " / %" PRIu64
                "   pi = %.8f"
                "   error = %.3e"
                "   elapsed = %.2f s\n",
                i,
                options.darts,
                pi_estimate,
                fabs(pi_estimate - PI_TRUE),
                elapsed
            );


            fflush(stdout);

            sleep_seconds(
                options.sleep_seconds
            );
        }


        /*
         * ----------------------------------------------------------
         * Periodic checkpoint
         * ----------------------------------------------------------
         */

        if (
            options.checkpoint_every > 0 &&
            i % options.checkpoint_every == 0
        ) {

            checkpoint_t checkpoint;

            initialize_checkpoint(
                &checkpoint
            );


            checkpoint.completed_darts = i;

            checkpoint.inside_circle =
                inside_circle;

            checkpoint.rng_state =
                rng_state;

            checkpoint.total_darts =
                options.darts;

            checkpoint.seed =
                options.seed;

            checkpoint.elapsed_seconds =
                previous_elapsed +
                wall_time() -
                start_time;


            save_checkpoint(
                options.checkpoint_file,
                &checkpoint
            );


            printf(
                "  --> checkpoint saved "
                "at dart %" PRIu64 "\n",
                i
            );

            fflush(stdout);
        }
    }


    const double total_elapsed =
        previous_elapsed +
        wall_time() -
        start_time;


    const double pi_estimate =
        4.0 *
        (double) inside_circle /
        (double) options.darts;


    /*
     * Final checkpoint.
     */

    checkpoint_t final_checkpoint;

    initialize_checkpoint(
        &final_checkpoint
    );

    final_checkpoint.completed_darts =
        options.darts;

    final_checkpoint.inside_circle =
        inside_circle;

    final_checkpoint.rng_state =
        rng_state;

    final_checkpoint.total_darts =
        options.darts;

    final_checkpoint.seed =
        options.seed;

    final_checkpoint.elapsed_seconds =
        total_elapsed;


    save_checkpoint(
        options.checkpoint_file,
        &final_checkpoint
    );


    printf("\n");
    printf("========================================================================\n");
    printf("Final result\n");
    printf("========================================================================\n");

    printf(
        "Darts thrown       : %" PRIu64 "\n",
        options.darts
    );

    printf(
        "Inside circle      : %" PRIu64 "\n",
        inside_circle
    );

    printf(
        "Estimated pi       : %.10f\n",
        pi_estimate
    );

    printf(
        "Reference pi       : %.10f\n",
        PI_TRUE
    );

    printf(
        "Absolute error     : %.6e\n",
        fabs(pi_estimate - PI_TRUE)
    );

    printf(
        "Total elapsed time : %.2f s\n",
        total_elapsed
    );

    printf("========================================================================\n");

    return EXIT_SUCCESS;
}
